packages/vitest/src/node/pools/threads.ts

```diff
@@ -147,13 +147,32 @@ export function createThreadsPool(ctx: Vitest, { execArgv, env }: PoolProcessOpt
 
       if (multipleThreads.length) {
```

Comment:
This and the `if (singleThreads.length) {` block below are now quite similar. I guess these could be merged into a single one..?

Reply:
They shouldn't be similar though, every worker should be a new one

Reply:
> every worker should be a new one

I think my earlier assumptions were incorrect. Are the statements below right?

- `--single-thread`
  - Workers should be recycled when environment options change
  - Workers should be recycled when workspace project changes
  - Multiple files should be run in a single worker
  - `--isolate true/false` should have no effect, since only one worker is used..? Should we actually raise an error when such condition is used, like we do with other conflicting options in `resolveConfig`?

- `--threads`
  - Workers should be recycled when environment options change
  - Workers should be recycled when workspace project changes
  - When `--isolate`, workers should be recycled between files
  - When `--no-isolate`, workers are not recycled between files

Reply:
> `--isolate true/false` should have no effect

It actually has an effect, because it doesn't reset the module graph and mock registry between tests

> `--threads`

I think the current implementation expects `--isolate` to be enabled. From what I see it's possible to have 4 workers created, but only one would be busy (but all 4 workers are already created), so it's kind of a waste. It would be good to have more things running in parallel (simplest example: 4 workspace projects with 1 test inside will run one test after another instead of 4 workers in parallel). I might be wrong here, feel free to correct me.

But yes, recycling options should be correct I think.

Reply:
> It would be good to have more things running in parallel (simplest example: 4 workspace projects with 1 test inside will run one test after another instead of 4 workers in parallel)

Right, we actually don't need any changes to the `--threads` + `--isolate` part. We can fill the pool with tasks as before and rely on `isolateWorkers`. Tests using different workspace projects and environments are run parallel as before.

But the case with `--threads` + `--no-isolate` part might be tricky, as we still need to isolate between workspace projects and environments. Easiest way to achieve this is to group all tests by environments and workspaces, and then run these test sets sequentially and call `pool.recycleWorkers` in between. But there may be idle workers waiting for next set's tests in this case.  

I'll need to check if we could fill the pool with all tasks and call `pool.recycleWorkers()` between the tasks when needed. Some on-going workers can then be running in different environments, but all new ones are started with their environments. But in this case the order of tasks has to be stable.

The `--single-thread` case seems simple and should be correct already in this PR.

Reply:
This is now implemented. When `--threads` + `--no-isolate`, the tasks are run parallel as efficient as possible. This is done by checking whether pool has any tasks in queue, and once the queue is empty, we can recycle workers and push more tasks to queue. In practice the pool is running different environment + workspace projects in parallel.